tests/components/fan/test_template.py

```diff
@@ -0,0 +1,536 @@
+        self._register_components()
+
+        # Turn on fan
+        components.fan.turn_on(self.hass, _TEST_FAN)
+        self.hass.block_till_done()
+
+        # Set fan's osc to True
+        components.fan.oscillate(self.hass, _TEST_FAN, True)
+        self.hass.block_till_done()
+
+        # verify
+        assert self.hass.states.get(_OSC_INPUT_SELECT).state == 'True'
+        self._verify(STATE_ON, None, True)
+
+        # Set fan's osc to False
+        components.fan.oscillate(self.hass, _TEST_FAN, False)
+        self.hass.block_till_done()
+
+        # verify
+        assert self.hass.states.get(_OSC_INPUT_SELECT).state == 'False'
+        self._verify(STATE_ON, None, False)
+
+    def test_set_invalid_osc_from_initial_state(self):
+        """Test set invalid oscillating when fan is in initial state."""
+        self._register_components()
+
+        # Turn on fan
+        components.fan.turn_on(self.hass, _TEST_FAN)
+        self.hass.block_till_done()
+
+        # Set fan's osc to 'invalid'
+        components.fan.oscillate(self.hass, _TEST_FAN, 'invalid')
+        self.hass.block_till_done()
+
+        # verify
+        assert self.hass.states.get(_OSC_INPUT_SELECT).state == ''
+        self._verify(STATE_ON, None, None)
+
+    def test_set_invalid_osc(self):
+        """Test set invalid oscillating when fan has valid osc."""
+        self._register_components()
+
+        # Turn on fan
+        components.fan.turn_on(self.hass, _TEST_FAN)
+        self.hass.block_till_done()
+
+        # Set fan's osc to True
+        components.fan.oscillate(self.hass, _TEST_FAN, True)
+        self.hass.block_till_done()
+
+        # verify
+        assert self.hass.states.get(_OSC_INPUT_SELECT).state == 'True'
+        self._verify(STATE_ON, None, True)
+
+        # Set fan's osc to False
+        components.fan.oscillate(self.hass, _TEST_FAN, None)
+        self.hass.block_till_done()
+
+        # verify osc is unchanged
+        assert self.hass.states.get(_OSC_INPUT_SELECT).state == 'True'
+        self._verify(STATE_ON, None, True)
+
+    def _verify(self, expected_state, expected_speed, expected_oscillating):
+        """Verify fan's state, speed and osc."""
+        state = self.hass.states.get(_TEST_FAN)
+        attributes = state.attributes
+        assert state.state == expected_state
+        assert attributes.get(ATTR_SPEED, None) == expected_speed
+        assert attributes.get(ATTR_OSCILLATING, None) == expected_oscillating
+
+    def _register_components(self, speed_list=None):
+        """Register basic components for testing."""
+        with assert_setup_component(1, 'input_boolean'):
+            assert setup.setup_component(
+                self.hass,
+                'input_boolean',
+                {'input_boolean': {'state': None}}
+            )
+
+        with assert_setup_component(2, 'input_select'):
+            assert setup.setup_component(self.hass, 'input_select', {
+                'input_select': {
```

Comment:
I mean lots of tests do this, but I don't think tests from one platform should depend on another integration, it's just bad testing (and makes changing those other integrations a pain).

Why not just use the state machine manually?

Reply:
I want them to be integration tests so we can sure that our scripts run successfully with real results. Please advise an example of state machine way of testing.